ocaml/xapi-storage/generator/lib/data.ml

```diff
@@ -116,6 +116,21 @@ module Datapath (R : RPC) = struct
       ]
       (dbg @-> uri_p @-> domain @-> returning unit error)
 
+  let activate_readonly =
+    declare "activate_readonly"
+      [
+        "[activate_readonly uri domain] is called just before a VM, or the "
+      ; "control domain, needs to read a volume. A single volume may be "
+      ; "activated readonly multiple times, including on multiple independent "
+      ; "hosts. It is not permitted for a volume to be activated both readonly "
+      ; "and read-write concurrently. Implementations shall declare the "
+      ; "ATTACH_READONLY feature for this method to be supported. This function "
+      ; "is idempotent and in all other respects is interchangeable with "
```

Comment:
What are the rules on moving from RO activation to RW activation? Should the volume first be deactivated RO and then activated RW, or would activating RW "upgrade" the virtual disk to RW?
Same question on going from RW to RO. (xapi has a state machine internally so it can probably accomodate either choice).

Also worth pointing out the scope of attach/activate: is it per host or per domain? Since you get 'domain' as a parameter here I assume that activate is per domain.
Same question for 'attach': is it host or domain scoped? I think we've got a bug here that XAPI (due to its SMAPIv1 state machine) assumes that attach is per host and reference counts it. But the new qemu code would very much like for this to be per-domain and causes localhost migration to fail when XAPI "optimizes away" the 2nd attach on the same host.

Reply:
Sure, I'll amend it, basically "deactivate all readonly and then you can activate 1 and only 1 read-write".

The attach issues are separate and a pre-existing API violation as the attach API is defined as taking the domain ID to which the attach is occurring so the "optimisation" was always incorrect and a violation of the API spec.